### PR TITLE
Feature/un 19 3 Adding CI & Deployment

### PR DESCRIPTION
**Pull request by kudlaj**
In this feature we add:

- CI/CD 
- k8 configuration
- configure linter

### CODE REVIEW

File: go/main.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	})
+
+	// Create an HTTP server with timeouts
+	server := &http.Server{
+		Addr:         ":5001",
+		Handler:      handler,
+		ReadTimeout:  5 * time.Second,
```
Comment (luispdm) [inline go/main.go:54]: Mental note for the future, not for now as this is just the first iteration: we need to set up `ReadHeaderTimeout` too

File: .golangci.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    - errcheck
+    - goconst
+    - gocritic
+    - gocyclo
+    - gofumpt
+    - goimports
+    - gosec
+    - gosimple
+    - govet
+    - ineffassign
```
Comment (luispdm) [inline .golangci.yml:46]: I think `revive` and `unparam` add value, we should add them. You can find the full list of supported linters here: https://golangci-lint.run/usage/linters/. `lll` would be nice for long lines but I leave that up to you.
If the linters raise warnings/errors that we can ignore for now, you can add a comment like this next to the offending line: `//nolint:$LINTER` (e.g. `//nolint:unparam`).

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    adduser -D nodeuser -u 1000 -G nodegroup
+
+WORKDIR /app
+COPY ./go .
+RUN go build -race -o universalnode .
+
+FROM alpine:3.18.4 AS final
+
+WORKDIR /app
+COPY --from=builder /app /app
```
Comment (luispdm) [inline Dockerfile]: I forgot to edit this in the UN-9 PR. Please change it to: `COPY --from=builder /app/universalnode .` so we only copy the binary

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+ENTRYPOINT ["./universalnode"]
+USER nobody:nobody
```
Comment (luispdm) [inline Dockerfile]: Add a new line at the end of the file please. Thank you

File: scripts/install-helmfile.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+
+# Enable the 'exit on error' option
+set -e
+
+# Install helm-diff plugin
+helm plugin install https://github.com/databus23/helm-diff --version v3.6.0
```
Comment (luispdm) [inline scripts/install-helmfile.sh]: Update helm-diff to v3.8.1 and helmfile to v0.157.0 please

File: .circleci/config.yml [hunk 6]
```diff
@@ -0,0 +1,126 @@
+version: 2.1
+
+orbs:
+  kube-orb: circleci/kubernetes@1.3.1
+  helm: circleci/helm@2.0.1
```
Comment (luispdm) [inline .circleci/config.yml]: Update to 3.0.0

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      - image: ipfs/kubo:v0.18.1
+    steps:
+      - checkout
+      - run:
+          name: Run go tests
+          command: |
+            cd go && go test -race ./...
+  dockerize:
+    docker:
+      - image: cimg/base:2022.12
```
Comment (luispdm) [inline .circleci/config.yml]: Update to 2023.09 here and in the rest of the file where cimg/base is used.
If you prefer, you can declare executors once and use them in the rest of the file. Example:
```yml
executors:
  cimg-base:
    docker:
      - image: cimg/base:2023.09
  another:
    docker:
      - image: ubuntu-2004:202201-02
# then use it:
dockerize:
  description: Something
  executor: cimg-base
```

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    steps:
+      - run:
+          name: Install helmfile
+          command: ./scripts/install-helmfile.sh
+
+jobs:
+  go_test:
+    docker:
+      - image: cimg/go:1.21
+      - image: ipfs/kubo:v0.18.1
```
Comment (luispdm) [inline .circleci/config.yml]: The kubo image is not required for now

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+              only: /.*/
+      - deploy_to_development:
+          filters:
+              branches:
+                only:
+                  - main
+          requires:
+            - lint
+            - go_test
+            - deployment_test
```
Comment (luispdm) [inline .circleci/config.yml]: Add `dockerize` (or `dockerize_universalnode`, I am not sure which of the two works, please look that up) and a new line at the end of the file, please

File: .circleci/config.yml [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+          stable-repo-url: https://charts.helm.sh/stable
+      - install_helmfile
+      - run:
+          name: Lint Helm charts for development environment
+          command: |
+            cd deployment/
+            TAG=$CIRCLE_SHA1 helmfile lint
+  deploy_to_development:
+    docker:
+      - image: cimg/deploy:2022.11
```
Comment (luispdm) [inline .circleci/config.yml]: Update to 2023.09 please

File: .circleci/config.yml [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    steps:
+      - checkout
+      - kube-orb/install-kubectl:
+          kubectl-version: "v1.16.12"
```
Comment (luispdm) [inline .circleci/config.yml]: I would suggest upgrading this to match the version of our cluster

File: .circleci/config.yml [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+  lint:
+    docker:
+      - image: golangci/golangci-lint:v1.54.2
+    steps:
+      - checkout
+      - run:
+          name: Run golangci-lint
+          command: |
+            cd go && golangci-lint run
+  deployment_test:
```
Comment (luispdm) [inline .circleci/config.yml]: Rename it to: `deployment_lint` pleae

File: deployment/charts/universalnode/Chart.yaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+appVersion: 1.16.0
```
Comment (luispdm) [inline deployment/charts/universalnode/Chart.yaml]: Change this to 0.1.0 as well please

File: deployment/charts/universalnode/templates/deployment.yaml [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+            name: rpc
+        resources:
+          requests:
+            memory: "512Mi"
+            cpu: "500m"
+          limits:
+            memory: "4096Mi"
+            cpu: "4"
+        command: [ "./universalnode" ]
+        args: []
```
Comment (luispdm) [inline deployment/charts/universalnode/templates/deployment.yaml]: Please remove `args` as it is not used for now

File: deployment/charts/universalnode/templates/deployment.yaml [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+        image: freeverseio/laos-universal-node:{{ .Values.imageTag }}
+        imagePullPolicy: IfNotPresent
+        ports:
+          - containerPort: 5001
+            name: rpc
+        resources:
+          requests:
+            memory: "512Mi"
+            cpu: "500m"
+          limits:
```
Comment (luispdm) [inline deployment/charts/universalnode/templates/deployment.yaml:27]: I suggest changing limits to `2Gi` for `memory` and `2` for `cpu`. It's better to start with low requirements IMHO

File: deployment/helmfile.yaml [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+releases:
+  - name: universalnode
+    namespace: laos-bigbang
+    chart: ./charts/universalnode
+    createNamespace: false
+    values:
+      - name: universalnode
+      - namespace: laos-bigbang
+      - imageTag: {{ requiredEnv "TAG" }}
+      - debug: true
```
Comment (luispdm) [inline deployment/helmfile.yaml]: Please remove `debug` as it is not used for now

File: deployment/charts/universalnode/templates/deployment.yaml [hunk 17]
```diff
@@ -0,0 +1,31 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ .Values.name }}
+  labels:
+    app.kubernetes.io/app: {{ .Values.name }}
+spec:
+  selector:
+    matchLabels:
+      app.kubernetes.io/app: universalnode
```
Comment (luispdm) [inline deployment/charts/universalnode/templates/deployment.yaml]: Please, replace `universalnode` here and on lines 14 and 17 with `{{ .Values.name }}`. Thanks

File: deployment/charts/universalnode/templates/service.yaml [hunk 18]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
```
Comment (luispdm) [inline deployment/charts/universalnode/templates/service.yaml:1]: Mental note: if we want the app to be reachable outside the cluster we have to add an ingress

**Review by luispdm — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 11 above)
Comment (kudlaj) [reply .circleci/config.yml]: how can we check it?

File: .golangci.yml (hunk 2 above)
Comment (kudlaj) [reply .golangci.yml:46]: I will leave it like it is for now

File: .circleci/config.yml (hunk 6 above)
Comment (kudlaj) [reply .circleci/config.yml]: I will not change this since it will bring a lot of problems.

File: .circleci/config.yml [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+          name: Run golangci-lint
+          command: |
+            cd go && golangci-lint run
+  deployment_lint:
+    docker:
+      - image: cimg/base:2023.09
+    steps:
+      - checkout
+      - helm/install-helm-client:
+          version: v3.11.0
```
Comment (luispdm) [inline .circleci/config.yml]: Ops. I forgot to mention that you can update this to install the latest version, v3.13.0 https://github.com/helm/helm/releases/tag/v3.13.0

Comment (luispdm) [reply .circleci/config.yml]: Resolving conversation as the upgrade has been performed

File: .circleci/config.yml (hunk 6 above)
Comment (luispdm) [reply .circleci/config.yml]: fixed https://github.com/freeverseio/laos-universal-node/pull/5/commits/def72d51d0e03e347e785132d2e9af4a371a7310, resolving the conversation

File: .circleci/config.yml (hunk 11 above)
Comment (luispdm) [reply .circleci/config.yml]: I cannot check the version of our cluster as I don't have that level of permission. We should ask the people with superpowers 💪

File: deployment/charts/universalnode/templates/deployment.yaml (hunk 15 above)
Comment (luispdm) [reply deployment/charts/universalnode/templates/deployment.yaml:27]: You've changed the memory in the requests and the CPU in the limits. I meant to change the values in the `limits` section only, sorry. Please change it to make it look like this:
```yml
requests:
  memory: "512Mi"
  cpu: "500m"
limits:
  memory: "2Gi"
  cpu: "2"
```

**Review by luispdm — APPROVED** [review]

**Review by AlexRipoll — APPROVED** [review]
looks good to me

**Pull request merged**